agent/src/agentclient.rs

```diff
@@ -312,119 +400,48 @@ impl<T: APIServerClient> BrupopAgent<T> {
-                        event!(Level::INFO, "Rebooting node to complete update");
-
-                        if running_desired_version(&bottlerocket_shadow_spec).await? {
-                            // previous execution `reboot` exited loop and did not update the custom resource
-                            // associated with this node. When re-enter loop and finished reboot,
-                            // try to update the custom resource.
-                            let update_node_status = match self
-                                .gather_system_metadata(bottlerocket_shadow_spec.state.clone())
-                                .await
-                            {
-                                Ok(update_node_status) => update_node_status,
-                                Err(_) => {
-                                    // Errors gathering system metadata are ignored (and also logged by `gather_system_metadata()`).
-                                    event!(Level::WARN, "An error occurred when gathering system metadata. Restarting event loop");
-                                    sleep(AGENT_SLEEP_DURATION).await;
-                                    continue;
-                                }
-                            };
-                            if self
-                                .update_metadata_custom_resource(update_node_status)
-                                .await
-                                .is_err()
-                            {
-                                // Errors updating BottlerocketShadow are ignored (and also logged by `update_metadata_custom_resource()`).
-                                event!(Level::WARN, "An error occurred when updating BottlerocketShadow. Restarting event loop");
-                                sleep(AGENT_SLEEP_DURATION).await;
-                                continue;
-                            };
-                        } else {
-                            boot_update()
-                                .await
-                                .context(agentclient_error::UpdateActions {
-                                    action: "Reboot".to_string(),
-                                })?;
+            match self.handle_update(&bottlerocket_shadow).await {
+                Ok(()) => {}
+                Err(e) => {
+                    match e {
+                        // Errors from update_status_in_custom_resource should be skipped and retry in next loop
+                        agentclient_error::Error::MissingBottlerocketShadowStatus
+                        | agentclient_error::Error::BottlerocketShadowStatusVersion { source: _ }
+                        | agentclient_error::Error::BottlerocketShadowStatusChosenUpdate {
+                            source: _,
+                        }
+                        | agentclient_error::Error::UpdateBottlerocketShadowResource {
+                            source: _,
```

Comment:
I'm worried about this list becoming stale. Can we reclassify errors from that function using a new `Error` type, that way we can just refer to the overarching type rather than having to disambiguate this way?

Reply:
I think I can create a new Error enum related to operate Shadow and put it under agentclient_error::Error. I will do that.

Reply:
Yeah, Same feeling here. And another concern would be if we have more errors needed to be handled in future, this  part has to be updated often and become lengthy. So that might be worth to pack it like you mentioned above create a new Error enum.